hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -0,0 +1,447 @@
+        aggIR.pType
+      }
+      case AggExplode(array, name, aggBody, isScan) => {
+        array.inferSetPType(env)
+
+        // TODO: how is this concretely different from bindEval?
+        // TODO: Needed? We currently do a much simpler type check in TypeCheck, no env
+        // also, what's the benefit of bindScan/agg over bindEval for type inference?
+        // or should this be a BindingEnv?
+        val boundEnv =
+          if (isScan)
+            env.bindScan(name, array.pType)
+          else
+            env.bindAgg(name, array.pType)
+
+        aggBody.inferSetPType(boundEnv)
+
+        aggBody.pType
+      }
+      case AggGroupBy(key, aggIR, isScan) => {
+        key.inferSetPType(env)
+        aggIR.inferSetPType(env)
+
+        PDict(key.pType, aggIR.pType)
+      }
+      case AggArrayPerElement(a, _, _, aggBody, _, _) => {
+        // TODO: Case "AggArrayPerElement" in Parser shows that aggBody already has its env from a,
+        // but that this is given as type
+        aggBody.inferSetPType(env)
+        PArray(aggBody.pType)
+      }
+      case ApplyAggOp(_, _, _, aggSig) => {
+        // TODO: Something else?
+        PType.canonical(AggOp.getType(aggSig))
+      }
+      case ApplyScanOp(_, _, _, aggSig) => {
+        // TODO: Something else?
+        PType.canonical(AggOp.getType(aggSig))
+      }
+      case MakeStruct(fields) => {
+        PStruct(fields.map {
+          case (name, a) => {
+            a.inferSetPType(env)
+            (name, a.pType)
+          }
+        }: _*)
+      }
+      case SelectFields(old, fields) => {
+        old.inferSetPType(env)
+        val tbs = coerce[PStruct](old.pType)
+        tbs.select(fields.toFastIndexedSeq)._1
+      }
+      case InsertFields(old, fields, fieldOrder) => {
+        old.inferSetPType(env)
+        val tbs = coerce[PStruct](old.pType)
+
+        // TODO: does this need old environment? No name
+        val s = tbs.insertFields(fields.map(f =>  {
+          f._2.inferSetPType(env)
+          (f._1, f._2.pType)
+        }))
+        fieldOrder.map { fds =>
+          assert(fds.length == s.size)
+          PStruct(fds.map(f => f -> s.fieldType(f)): _*)
+        }.getOrElse(s)
+      }
+      case GetField(o, name) => {
+        o.inferSetPType(env)
+        val t = coerce[PStruct](o.pType)
+        if (t.index(name).isEmpty)
+          throw new RuntimeException(s"$name not in $t")
+        // TODO: Why is this not a copy?
+        val fd = t.field(name).typ
+        fd.setRequired(t.required && fd.required)
+      }
+      case MakeTuple(values) => {
+        PTuple(values.map(v => {
+          v.inferSetPType(env)
+          v.pType
+        }).toFastIndexedSeq)
+      }
+      case GetTupleElement(o, idx) => {
+        o.inferSetPType(env)
+        val t = coerce[PTuple](o.pType)
+        assert(idx >= 0 && idx < t.size)
+        val fd = t.types(idx)
+        fd.setRequired(t.required && fd.required)
+      }
+      case TableCount(_) => {
```

Comment:
Everything from here down to `CollectDistributedArray` doesn't need a PType, I think -- we don't have code-generators for them, so they can't appear in Compile

Reply:
To check: I think you meant half-closed range, inclusive of TableCount and exclusive of CollectDistributedArray: I checked Emit, which I believe defines all IR nodes that we generate code for, CollectDistributedArray is present.